### DISCUSSION TITLE
Comparing growth of an organism in monoculture vs. coculture

### DISCUSSION THREAD

**Discussion by justuswfink** [micom-dev/micom#147] [Q&A] [answered]
Dear MICOM team, dear community,

we are interested in the growth of a focal taxon in a mixed community and have been using the 'cooperative tradeoff' for the community optimization. Great to have this package!

Now we want to compare the metabolic fluxes of our focal strain to the fluxes in monoculture to see what new nutrients are taken up. However, we are unsure if we should use the 'classic FBA' or the 'cooperative tradeoff' for the monoculture. Do you have a recommendation? 

Here is a pseudocode of what we are currently doing:

`taxa  =  a dataframe of metabolic models`

`# compute fluxes in community using cooperative trade-off`
`com = Community(taxa) `
`sol_com = com.cooperative_tradeoff(fluxes=True)`

`# compute fluxes in monoculture using cooperative trade-off`
`mono = Community(taxa.loc[i]) # index i is our focal strain`
`sol_mono = mono.cooperative_tradeoff(fluxes=True)`

As a second approach for the monoculture, we also considered 'standard FBA':

`sol_mono_fba = mono.optimize(fluxes = True)`

The fluxes for the monoculture from cooperative trade-off do in general not match the fluxes from FBA. The plot shows many more nutrients being taken up in cooperative tradeoff ( `sol_mono.fluxes`, y-axis) compared to standard FBA (`sol_mono_fba.fluxes`,x-axis). 

![Micom_FBA_vs Micom_Cooperative_Tradeoff](https://github.com/micom-dev/micom/assets/144257498/a73e8a12-8b04-403b-bd30-3f14f5243879)

We were surprised to see this disagreement. Intuitively, we were thinking that the optimization problem for 'cooperative tradeoff' in a single species should exactly match the optimization problem solved in 'standard FBA'. Is this a correct interpretation of the 'cooperative tradeoff' framework?

We are happy to look into more references and also learn about other ways for comparing monoculture to coculture with MICOM!

Justus

**Comment by cdiener** [comment, marked as answer]
Hi, sorry for the late reply. You are correct in that a single taxon cooperative tradeoff (with fraction=1) is the same as a normal FBA. 

I think this is because `Community.optimize(fluxes=True)` uses pFBA by default, whereas `Community.cooperative_tradeoff(fluxes=True)` does not. If you run your code with `sol_mono = mono.cooperative_tradeoff(fluxes=True, pfba=True)` and `sol_mono_fba = mono.optimize(fluxes=True, pfba=True)` it should give you similar results. I will fix the default arguments so they match in the next release.

**Reply by justuswfink** [reply]
Hi Christian, thank you very much for pointing me to these different defaults! I checked again with our code, and this removes 99% of the disagreement. I will post a quick summary below, also for future users.

**Comment by justuswfink** [comment]
This is a quick demonstration of how the pFBA=True setting helps to align the two optimization methods (classic FBA and Cooperative Trade-Off), with some small disagreement remaining.

To reproduce the initial behavior, you can use the standard E.coli model (iML1515) and create a monoculture using the Community object in MICOM as follows. Throughout this example, I am using the 'Gurobi' solver. 

```python
# create a datatable for MICOM 
taxa = pd.DataFrame(columns = ['id', 'file'])
taxa.loc[0] = pd.Series({'id': 'iML1515', 'file' : 'iML1515.xml'})
# construct MICOM Community object
com = Community(all_taxa, solver = 'gurobi' )
# we explicit choose tolerance parameters, to make sure they agree
solver_params = {'atol': 1e-12, 'rtol': 1e-8}

# get solution using the two methods available in MICOM
sol_fba = com.optimize(fluxes=True,**solver_params) ## using classic FBA
sol_coop = com.cooperative_tradeoff(fluxes=True, fraction = 1.0, **solver_params) ## using the cooperative trade-off method
```
The scatter plot shows large disagreement in the flux values, which gave rise to this thread in the first place.
![ecoli_model_pfba_default](https://github.com/micom-dev/micom/assets/144257498/b23ae0bf-671d-4ba4-9f18-353894ad2874)


This disagreement almost vanishes, when we actively enforce the use of parsimonious FBA to determine the fluxes (pFBA=True). The two lines to adjust here are: 

```python
sol_fba = com.optimize(fluxes=True, pfba=True, **solver_params) ## using classic FBA
sol_coop = com.cooperative_tradeoff(fluxes=True,fraction = 1.0, pfba=True, **solver_params) ## using the cooperative trade-off method

```
![ecoli_model_pbfa_true](https://github.com/micom-dev/micom/assets/144257498/88e4119d-5ebd-464f-8712-ad6db0e0a45a)

The scatter plot (left panel) shows excellent alignment over the entire range of flux values, however, there is still some remaining discrepancy when we look at the residuals (right side). The residual plot shows that a few individual flux values still differ by an absolute value of 10^-2.  These flux values happen to be internal. A look into the underlying datatable shows that the Cooperative Tradeoff chooses to use a different transport reaction (`PROt2rpp`) than the classic FBA (`GLYCLTt2rpp`). 

<img width="723" alt="Screenshot 2023-12-23 at 20 41 47" src="https://github.com/micom-dev/micom/assets/144257498/608fa36d-49d5-4941-a8a2-87b8445cc0fa">

I am not sure what this remaining discrepancy means and where it comes form. Importantly, it only seems to affect internal reactions. When we only look at exchange fluxes at the model boundary (labeled with suffix `_m` in MICOM), the disagreement is below 10^-7. 

My take-aways from this question are: 
- The flux distribution from the MICOM Cooperative Tradeoff strategy agrees with classic FBA, as long as one explicitly activates parsimonious FBA with `pFBA=True`. 
- Some disagreement remains and it cannot be explained by numerical error, since the 10^-2 disagreement is much larger than the numerical error typically observed with the Gurobi solver (<10^-8). 
- At least the fluxes from the growth medium into the communty (`_m` suffix) are robustly identified by both methods. 

Hope this helps, I am happy to share more code if this comes up in the future for any other user. All the best, 

Justus

**Reply by justuswfink** [reply]
Ok, I want to add the following to my summary. In hind-sight, the remaining differences here are to be expected, simply because the two attempts to solve the same optimization proble happen to return two different solutions from the total solution space. The table shows that the difference here consists  in the presence or absence of pairs of internal fluxes, that happen to cycle nutrients. 

In this specific case, the two fluxes starting with `GLYCLT...` efffectively cycle 'glycolate' between periplasm and cytoplasm and the two fluxes starting with `PRO...` cycle the 'proline' across the same layers. Apparently, these cycles can always appear in an FBA solution for the internal reactions, but will not appear in the exchange reactions at the boundary of the model (since there is only one exchange reaction per compound). Still learning!
